### PR TITLE
[graph_trainer] Integrate precompile into AOT compilation path

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #2646
* #2650
* __->__ #2644
* #2643
* #2642

Add precompile and precompile_artifact_dir config fields to
GraphTrainerCompileConfig, and wire serializable compilation through
joint_graph_builder and the AOT compile dispatcher:

- joint_graph_builder: add serializable and on_compile params. When
  serializable=True, passes it to aot_compile_joint_with_descriptors.
  Calls on_compile callback after compilation with the compiled fn and
  tree specs.

- CompiledModule: add precompiled_fn param. When set, uses the
  precompiled function directly instead of calling joint_graph_builder.

- _apply_aot_compile: when precompile=True, checks for existing
  artifacts and loads them (skipping compilation). Otherwise compiles
  with serializable=True and saves the artifact via on_compile callback.

- _make_precompile_callback: builds the on_compile callback, validates
  that full_inductor_compilation is in the pass pipeline (required for
  serializable compilation).

Config fingerprint is threaded through the save and load paths:
fingerprint is computed in _make_precompile_callback (save path) and in
_apply_aot_compile (load path). On fingerprint mismatch, the stale
artifact is automatically discarded and recompilation proceeds — users
just pass --precompile and the right thing happens.

@claude

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
## Loss Curve Comparison: Precompile vs Baseline

Three scenarios were tested on 8x H100 GPUs with `--debug.deterministic --debug.seed=42` for 50 steps:

1. **Baseline** — `graph_trainer_llama3_debugmodel` (default compile, no AOT)
2. **AOT precompile save** — `graph_trainer_llama3_precompile` (AOT compile, saves artifact)
3. **AOT precompile load** — same config (loads saved artifact)

### Results

| Comparison | Max |diff| | Avg |diff| | Bit-identical? |
|---|---|---|---|
| AOT precompile vs baseline | 8.06e-04 | 1.39e-04 | No (different compile modes) |
| **Precompile save vs load** | **0.00** | **0.00** | **Yes** |

Key findings:
- AOT precompile and baseline produce nearly identical losses (tiny numerical differences expected from different compilation strategies)
- **Precompile save and load produce bit-identical losses**, confirming the serialization/deserialization roundtrip is exact
- Both loss curves converge from ~8.1 to ~2.85 over 50 steps

### Sample losses (first 10 steps)

```
Step   Baseline         Precompile       Diff
1      8.1373796463     8.1378803253     +0.0005006790
2      7.8131389618     7.8123331070     -0.0008058548
3      7.0552430153     7.0551223755     -0.0001206398
4      6.1516456604     6.1519870758     +0.0003414154
5      5.1983757019     5.1987795830     +0.0004038811
6      4.5588293076     4.5588674545     +0.0000381470
7      4.1592912674     4.1593523026     +0.0000610352
8      3.9065964222     3.9064943790     -0.0001020432
9      3.8438191414     3.8435192108     -0.0002999306
10     3.5995364189     3.5992548466     -0.0002815723
```

File: torchtitan/experiments/graph_trainer/compile.py [hunk 1]
```diff
@@ -96,15 +177,53 @@ def _apply_aot_compile(
...
+    model: nn.Module,
+    parallel_dims: ParallelDims,
+    storage: StorageAdapter,
+    artifact_key: str,
+    config_fingerprint: str = "",
+) -> CompiledModule:
+    """Load a precompiled artifact and wrap the model with it."""
+    from torchtitan.experiments.graph_trainer.precompile import precompile_load
+
+    register_blockmask_pytree_node()
```
Comment (bobrenjc93) [inline torchtitan/experiments/graph_trainer/compile.py:206]: can you add a comment here?
